File: python/src/fluent_cards/builders/action_builder.py
from __future__ import annotations
from typing import Optional, Union
from ..enums import ActionStyle, AssociatedInputs, ActionMode


class ActionBuilder:
    """Fluent builder for creating Action elements."""

    def __init__(self):
        self._action: Optional[dict] = None
        self._data_set: bool = False
        self._teams_data_set: bool = False
        self._teams_submit_typed_set: bool = False
        self._teams_submit_raw_set: bool = False

    def open_url(self, url: str, title: Optional[str] = None) -> ActionBuilder:
        """Sets the action type to Action.OpenUrl and specifies the URL to open.

        Args:
            url: The URL to open when the action is invoked.
            title: Optional title displayed on the action button.

        Returns:
            The builder instance for method chaining.
        """
        self._action = {'type': 'Action.OpenUrl', 'url': url, 'title': title}
        return self

    def submit(self, title: Optional[str] = None) -> ActionBuilder:
        """Sets the action type to Action.Submit.

        Args:
            title: Optional title displayed on the action button.

        Returns:
            The builder instance for method chaining.
        """
        self._action = {'type': 'Action.Submit', 'title': title}
        return self

    def show_card(self, title: Optional[str] = None) -> ActionBuilder:
        """Sets the action type to Action.ShowCard.

        Args:
            title: Optional title displayed on the action button.

        Returns:
            The builder instance for method chaining.
        """
        self._action = {'type': 'Action.ShowCard', 'title': title}
        return self

    def toggle_visibility(self, title: Optional[str] = None) -> ActionBuilder:
        """Sets the action type to Action.ToggleVisibility.

        Args:
            title: Optional title displayed on the action button.

        Returns:
            The builder instance for method chaining.
        """
        self._action = {'type': 'Action.ToggleVisibility', 'title': title}
        return self

    def execute(self, title: Optional[str] = None) -> ActionBuilder:
        """Sets the action type to Action.Execute.

        Args:
            title: Optional title displayed on the action button.

        Returns:
            The builder instance for method chaining.
        """
        self._action = {'type': 'Action.Execute', 'title': title}
        return self

    def _ensure_action_type_set(self) -> None:
        if self._action is None:
            raise ValueError(
                'No action type specified. Call open_url(), submit(), show_card(), '
                'toggle_visibility(), or execute() before setting properties.'
            )

    def with_id(self, id: str) -> ActionBuilder:
        """Sets the unique identifier for the action.

        Args:
            id: The unique identifier.

        Returns:
            The builder instance for method chaining.
        """
        self._ensure_action_type_set()
        self._action['id'] = id
        return self

    def with_title(self, title: str) -> ActionBuilder:
        """Sets the title displayed on the action button.

        Args:
            title: The button title text.

        Returns:
            The builder instance for method chaining.
        """
        self._ensure_action_type_set()
        self._action['title'] = title
        return self

    def with_icon_url(self, icon_url: str) -> ActionBuilder:
        """Sets the icon URL displayed on the action button.

        Args:
            icon_url: The URL of the icon image.

        Returns:
            The builder instance for method chaining.
        """
        self._ensure_action_type_set()
        self._action['iconUrl'] = icon_url
        return self

    def with_style(self, style: ActionStyle) -> ActionBuilder:
        """Sets the visual style of the action button.

        Args:
            style: The action style.

        Returns:
            The builder instance for method chaining.
        """
        self._ensure_action_type_set()
        self._action['style'] = style.value
        return self

    def with_is_enabled(self, is_enabled: bool) -> ActionBuilder:
        """Sets whether the action button is enabled.

        Args:
            is_enabled: False to disable the button.

        Returns:
            The builder instance for method chaining.
        """
        self._ensure_action_type_set()
        self._action['isEnabled'] = is_enabled
        return self

    def with_tooltip(self, tooltip: str) -> ActionBuilder:
        """Sets the tooltip text for the action button.

        Args:
            tooltip: The tooltip text to display on hover.

        Returns:
            The builder instance for method chaining.
        """
        self._ensure_action_type_set()
        self._action['tooltip'] = tooltip
        return self

    def with_data(self, data) -> ActionBuilder:
        """Sets the data payload for Action.Submit or Action.Execute.

        Args:
            data: The data to submit when the action is invoked.

        Returns:
            The builder instance for method chaining.

        Raises:
            ValueError: If not a Submit or Execute action, or if with_teams_data
                was already called.
        """
        self._ensure_submit_or_execute('with_data')
        if self._teams_data_set:
            raise ValueError(
                'Cannot use both with_data and with_teams_data on the same action. '
                'Use with_teams_data to combine msteams properties with custom data, '
                'or with_data for raw data.'
            )
        self._action['data'] = data
        self._data_set = True
        return self

    def with_associated_inputs(self, associated_inputs: AssociatedInputs) -> ActionBuilder:
        """Sets which inputs are submitted with this Action.Submit or Action.Execute.

        Args:
            associated_inputs: The associated inputs policy.

        Returns:
            The builder instance for method chaining.

        Raises:
            ValueError: If not a Submit or Execute action.
        """
        self._ensure_submit_or_execute('with_associated_inputs')
        self._action['associatedInputs'] = associated_inputs.value
        return self

    def with_verb(self, verb: str) -> ActionBuilder:
        """Sets the verb for an Action.Execute action.

        Args:
            verb: The verb string sent to the bot.

        Returns:
            The builder instance for method chaining.

        Raises:
            ValueError: If not an Execute action.
        """
        self._ensure_execute_only('with_verb')
        self._action['verb'] = verb
        return self

    def with_card(self, card: dict) -> ActionBuilder:
        """Sets the inline card for an Action.ShowCard action.

        Args:
            card: The Adaptive Card dictionary to show.

        Returns:
            The builder instance for method chaining.
        """
        self._ensure_action_type_set()
        if self._action.get('type') == 'Action.ShowCard':
            self._action['card'] = card
        return self

    def add_target_element(self, element_id: str, is_visible: Optional[bool] = None) -> ActionBuilder:
        """Adds a target element for an Action.ToggleVisibility action.

        Args:
            element_id: The ID of the element to toggle.
            is_visible: Optional explicit visibility state to set. If None, toggles the current state.

        Returns:
            The builder instance for method chaining.
        """
        self._ensure_action_type_set()
        if self._action.get('type') == 'Action.ToggleVisibility':
            if 'targetElements' not in self._action:
                self._action['targetElements'] = []
            if is_visible is None:
                self._action['targetElements'].append(element_id)
            else:
                self._action['targetElements'].append({'elementId': element_id, 'isVisible': is_visible})
        return self

    def with_mode(self, mode: ActionMode) -> ActionBuilder:
        """Sets whether the action is primary or secondary."""
        self._ensure_action_type_set()
        self._action['mode'] = mode.value
        return self

    def with_requires(self, key: str, version: str) -> ActionBuilder:
        """Sets a feature requirement for the action."""
        self._ensure_action_type_set()
        if 'requires' not in self._action:
            self._action['requires'] = {}
        self._action['requires'][key] = version
        return self

    def with_fallback(self, fallback) -> ActionBuilder:
        """Sets the fallback content if the action is unsupported."""
        self._ensure_action_type_set()
        self._action['fallback'] = fallback
        return self

    # ── Teams-specific methods (Submit-only) ────────────────────────────────

    def _ensure_submit_or_execute(self, method_name: str) -> None:
        self._ensure_action_type_set()
        if self._action.get('type') not in ('Action.Submit', 'Action.Execute'):
            raise ValueError(
                f'{method_name}() is only available on Submit or Execute actions. '
                'Call submit() or execute() before using this method.'
            )

    def _ensure_execute_only(self, method_name: str) -> None:
        self._ensure_action_type_set()
        if self._action.get('type') != 'Action.Execute':
            raise ValueError(
                f'{method_name}() is only available on Execute actions. '
                'Call execute() before using this method.'
            )

    def _ensure_submit_only(self, method_name: str) -> None:
        self._ensure_action_type_set()
        if self._action.get('type') != 'Action.Submit':
            raise ValueError(
                f'{method_name} is only available on Submit actions. '
                'Call submit() before using this method.'
            )

    def _ensure_no_data_conflict(self) -> None:
        if self._data_set:
            raise ValueError(
                'Cannot use both with_data and with_teams_data on the same action. '
                'Use with_teams_data to combine msteams properties with custom data, '
                'or with_data for raw data.'
            )

    def with_teams_task_fetch(self) -> ActionBuilder:
        """Sets the action data to ``{'msteams': {'type': 'task/fetch'}}`` (Submit-only).

        Returns:
            The builder instance for method chaining.

        Raises:
            ValueError: If not a Submit action or if with_data was already called.
        """
        from .teams_data_builder import TeamsDataBuilder
        self._ensure_submit_only('with_teams_task_fetch')
        self._ensure_no_data_conflict()
        b = TeamsDataBuilder()
        b.with_task_fetch()
        self._action['data'] = b.build()
        self._teams_data_set = True
        return self

    def with_teams_data(self, configure) -> ActionBuilder:
        """Configures a Teams-specific data payload (Submit-only).

        Args:
            configure: A callable that receives a TeamsDataBuilder.

        Returns:
            The builder instance for method chaining.

        Raises:
            ValueError: If not a Submit action or if with_data was already called.
        """
        from .teams_data_builder import TeamsDataBuilder
        self._ensure_submit_only('with_teams_data')
        self._ensure_no_data_conflict()
        b = TeamsDataBuilder()
        configure(b)
        self._action['data'] = b.build()
        self._teams_data_set = True
        return self

    def with_teams_submit_feedback(self, configure) -> ActionBuilder:
        """Configures Teams submit feedback properties (Submit-only).

        Args:
            configure: A callable that receives a TeamsSubmitPropertiesBuilder.

        Returns:
            The builder instance for method chaining.

        Raises:
            ValueError: If not a Submit action or if with_teams_submit_raw was called.
        """
        from .teams_submit_properties_builder import TeamsSubmitPropertiesBuilder
        self._ensure_submit_only('with_teams_submit_feedback')
        if self._teams_submit_raw_set:
            raise ValueError(
                'Cannot use both with_teams_submit_feedback and with_teams_submit_raw '
                'on the same action. Use one or the other.'
            )
        b = TeamsSubmitPropertiesBuilder()
        configure(b)
        self._action['msteams'] = b.build()
        self._teams_submit_typed_set = True
        return self

    def with_teams_submit_raw(self, value: dict) -> ActionBuilder:
        """Sets the Teams action-level msteams property from a raw dict (Submit-only).

        Args:
            value: The raw msteams properties dictionary.

        Returns:
            The builder instance for method chaining.

        Raises:
            ValueError: If not a Submit action or if with_teams_submit_feedback was called.
        """
        self._ensure_submit_only('with_teams_submit_raw')
        if self._teams_submit_typed_set:
            raise ValueError(
                'Cannot use both with_teams_submit_feedback and with_teams_submit_raw '
                'on the same action. Use one or the other.'
            )
        self._action['msteams'] = dict(value)
        self._teams_submit_raw_set = True
        return self

    def build(self) -> dict:
        """Builds and returns the configured action dictionary.

        Returns:
            The configured action dictionary.

        Raises:
            ValueError: If no action type has been set.
        """
        if self._action is None:
            raise ValueError(
                'No action type specified. Call open_url(), submit(), show_card(), '
                'toggle_visibility(), or execute() first.'
            )
        return self._action

File: node/packages/fluent-cards/src/builders/ActionBuilder.ts
import type {
  AdaptiveAction,
  OpenUrlAction,
  SubmitAction,
  ShowCardAction,
  ToggleVisibilityAction,
  ExecuteAction,
  TargetElement,
  AdaptiveCard,
} from '../models.js';
import { ActionMode, ActionStyle, AssociatedInputs } from '../enums.js';
import { TeamsDataBuilder } from './TeamsDataBuilder.js';
import { TeamsSubmitPropertiesBuilder } from './TeamsSubmitPropertiesBuilder.js';

/**
 * Fluent builder for {@link AdaptiveAction} instances.
 *
 * Call one of the action-type methods first (`openUrl`, `submit`, etc.)
 * then chain modifiers, then call `build()`.
 */
export class ActionBuilder {
  private action: AdaptiveAction | undefined;
  private dataSet = false;
  private teamsDataSet = false;
  private teamsSubmitTypedSet = false;
  private teamsSubmitRawSet = false;

  /** Sets the action type to OpenUrl. @param url The URL to open. @param title Optional button title. @returns The builder instance for method chaining. */
  openUrl(url: string, title?: string): this {
    this.action = { type: 'Action.OpenUrl', url, title } as OpenUrlAction;
    return this;
  }

  /** Sets the action type to Submit. @param title Optional button title. @returns The builder instance for method chaining. */
  submit(title?: string): this {
    this.action = { type: 'Action.Submit', title } as SubmitAction;
    return this;
  }

  /** Sets the action type to ShowCard. @param title Optional button title. @returns The builder instance for method chaining. */
  showCard(title?: string): this {
    this.action = { type: 'Action.ShowCard', title } as ShowCardAction;
    return this;
  }

  /** Sets the action type to ToggleVisibility. @param title Optional button title. @returns The builder instance for method chaining. */
  toggleVisibility(title?: string): this {
    this.action = { type: 'Action.ToggleVisibility', title } as ToggleVisibilityAction;
    return this;
  }

  /** Sets the action type to Execute (Universal Action Model). @param title Optional button title. @returns The builder instance for method chaining. */
  execute(title?: string): this {
    this.action = { type: 'Action.Execute', title } as ExecuteAction;
    return this;
  }

  /** Ensures an action type has been set before modifying properties. @throws Error if no action type has been selected. */
  private ensureActionTypeSet(): void {
    if (!this.action) {
      throw new Error('No action type has been specified. Call openUrl(), submit(), showCard(), toggleVisibility(), or execute() before setting properties.');
    }
  }

  /** Sets the unique identifier for the action. @param id The unique identifier. @returns The builder instance for method chaining. */
  withId(id: string): this {
    this.ensureActionTypeSet();
    this.action!.id = id;
    return this;
  }

  /** Sets the button title. @param title The button label text. @returns The builder instance for method chaining. */
  withTitle(title: string): this {
    this.ensureActionTypeSet();
    this.action!.title = title;
    return this;
  }

  /** Sets the icon URL displayed on the button. @param iconUrl The icon URL. @returns The builder instance for method chaining. */
  withIconUrl(iconUrl: string): this {
    this.ensureActionTypeSet();
    this.action!.iconUrl = iconUrl;
    return this;
  }

  /** Sets the visual style of the action button. @param style The action style. @returns The builder instance for method chaining. */
  withStyle(style: ActionStyle): this {
    this.ensureActionTypeSet();
    this.action!.style = style;
    return this;
  }

  /** Sets whether the action is enabled. @param isEnabled True to enable the action. @returns The builder instance for method chaining. */
  withIsEnabled(isEnabled: boolean): this {
    this.ensureActionTypeSet();
    this.action!.isEnabled = isEnabled;
    return this;
  }

  /** Sets the tooltip shown on hover. @param tooltip The tooltip text. @returns The builder instance for method chaining. */
  withTooltip(tooltip: string): this {
    this.ensureActionTypeSet();
    this.action!.tooltip = tooltip;
    return this;
  }

  /** Sets the action mode (primary or secondary). @param mode The action mode. @returns The builder instance for method chaining. */
  withMode(mode: ActionMode): this {
    this.ensureActionTypeSet();
    this.action!.mode = mode;
    return this;
  }

  /** Sets the feature requirements for the action. @param key The feature name. @param version The minimum required version. @returns The builder instance for method chaining. */
  withRequires(key: string, version: string): this {
    this.ensureActionTypeSet();
    this.action!.requires = { ...this.action!.requires, [key]: version };
    return this;
  }

  /** Sets the fallback behavior when the action is unsupported. @param fallback The fallback value ('drop' or an action). @returns The builder instance for method chaining. */
  withFallback(fallback: 'drop' | AdaptiveAction): this {
    this.ensureActionTypeSet();
    this.action!.fallback = fallback;
    return this;
  }

  // ─── SubmitAction / ExecuteAction specifics ───────────────────────────────

  /** Sets the data payload (for Submit or Execute actions). @param data The data to submit. @returns The builder instance for method chaining. @throws Error if not a Submit or Execute action, or if WithTeamsData or WithTeamsTaskFetch was already called. */
  withData(data: unknown): this {
    this.ensureActionTypeSet();
    this.ensureSubmitOrExecute('withData');
    if (this.teamsDataSet) {
      throw new Error('Cannot use both withData and withTeamsData on the same action. Use withTeamsData to combine msteams properties with custom data, or withData for raw data.');
    }
    (this.action as SubmitAction | ExecuteAction).data = data;
    this.dataSet = true;
    return this;
  }

  /** Sets which inputs are included when the action is submitted (for Submit or Execute actions). @param associatedInputs The associated inputs option. @returns The builder instance for method chaining. @throws Error if not a Submit or Execute action. */
  withAssociatedInputs(associatedInputs: AssociatedInputs): this {
    this.ensureActionTypeSet();
    this.ensureSubmitOrExecute('withAssociatedInputs');
    (this.action as SubmitAction | ExecuteAction).associatedInputs = associatedInputs;
    return this;
  }

  // ─── ExecuteAction specific ───────────────────────────────────────────────

  /** Sets the verb for an Execute action. @param verb The command verb. @returns The builder instance for method chaining. @throws Error if not an Execute action. */
  withVerb(verb: string): this {
    this.ensureActionTypeSet();
    this.ensureExecuteOnly('withVerb');
    (this.action as ExecuteAction).verb = verb;
    return this;
  }

  // ─── ShowCardAction specific ──────────────────────────────────────────────

  /** Sets the card to reveal for a ShowCard action. @param card The embedded AdaptiveCard. @returns The builder instance for method chaining. */
  withCard(card: AdaptiveCard): this {
    this.ensureActionTypeSet();
    if (this.action!.type === 'Action.ShowCard') {
      (this.action as ShowCardAction).card = card;
    }
    return this;
  }

  // ─── ToggleVisibilityAction specific ─────────────────────────────────────

  /** Adds a target element to toggle for a ToggleVisibility action. @param elementId The ID of the element to toggle. @param isVisible Optional forced visibility state. @returns The builder instance for method chaining. */
  addTargetElement(elementId: string, isVisible?: boolean): this {
    this.ensureActionTypeSet();
    if (this.action!.type === 'Action.ToggleVisibility') {
      const a = this.action as ToggleVisibilityAction;
      a.targetElements ??= [];
      if (isVisible === undefined) {
        a.targetElements.push(elementId);
      } else {
        a.targetElements.push({ elementId, isVisible });
      }
    }
    return this;
  }

  // ─── Teams-specific methods (Submit-only) ────────────────────────────────

  /** Sets the action data to `{ msteams: { type: 'task/fetch' } }` for Submit actions. @returns The builder instance for method chaining. @throws Error if not a Submit action or if withData was already called. */
  withTeamsTaskFetch(): this {
    this.ensureSubmitOnly('withTeamsTaskFetch');
    this.ensureNoDataConflict();
    const b = new TeamsDataBuilder();
    b.withTaskFetch();
    (this.action as SubmitAction).data = b.build();
    this.teamsDataSet = true;
    return this;
  }

  /** Configures a Teams-specific data payload with msteams properties and/or custom data. Only available on Submit actions. @param configure A callback to configure the TeamsDataBuilder. @returns The builder instance for method chaining. @throws Error if not a Submit action or if withData was already called. */
  withTeamsData(configure: (b: TeamsDataBuilder) => void): this {
    this.ensureSubmitOnly('withTeamsData');
    this.ensureNoDataConflict();
    const b = new TeamsDataBuilder();
    configure(b);
    (this.action as SubmitAction).data = b.build();
    this.teamsDataSet = true;
    return this;
  }

  /** Configures Teams submit feedback properties (e.g. hide feedback). Only available on Submit actions. @param configure A callback to configure the TeamsSubmitPropertiesBuilder. @returns The builder instance for method chaining. @throws Error if not a Submit action or if withTeamsSubmitRaw was already called. */
  withTeamsSubmitFeedback(configure: (b: TeamsSubmitPropertiesBuilder) => void): this {
    this.ensureSubmitOnly('withTeamsSubmitFeedback');
    if (this.teamsSubmitRawSet) {
      throw new Error('Cannot use both withTeamsSubmitFeedback and withTeamsSubmitRaw on the same action. Use one or the other.');
    }
    const b = new TeamsSubmitPropertiesBuilder();
    configure(b);
    (this.action as SubmitAction).msteams = b.build();
    this.teamsSubmitTypedSet = true;
    return this;
  }

  /** Sets the Teams action-level msteams property from a raw object (escape hatch). Only available on Submit actions. @param value The raw msteams properties object. @returns The builder instance for method chaining. @throws Error if not a Submit action or if withTeamsSubmitFeedback was already called. */
  withTeamsSubmitRaw(value: Record<string, unknown>): this {
    this.ensureSubmitOnly('withTeamsSubmitRaw');
    if (this.teamsSubmitTypedSet) {
      throw new Error('Cannot use both withTeamsSubmitFeedback and withTeamsSubmitRaw on the same action. Use one or the other.');
    }
    (this.action as SubmitAction).msteams = { ...value };
    this.teamsSubmitRawSet = true;
    return this;
  }

  private ensureSubmitOrExecute(methodName: string): void {
    if (this.action!.type !== 'Action.Submit' && this.action!.type !== 'Action.Execute') {
      throw new Error(`${methodName}() is only available on Submit or Execute actions. Call submit() or execute() before using this method.`);
    }
  }

  private ensureExecuteOnly(methodName: string): void {
    if (this.action!.type !== 'Action.Execute') {
      throw new Error(`${methodName}() is only available on Execute actions. Call execute() before using this method.`);
    }
  }

  private ensureSubmitOnly(methodName: string): void {
    this.ensureActionTypeSet();
    if (this.action!.type !== 'Action.Submit') {
      throw new Error(`${methodName} is only available on Submit actions. Call submit() before using this method.`);
    }
  }

  private ensureNoDataConflict(): void {
    if (this.dataSet) {
      throw new Error('Cannot use both withData and withTeamsData on the same action. Use withTeamsData to combine msteams properties with custom data, or withData for raw data.');
    }
  }

  /** Builds and returns the configured action. @returns The configured AdaptiveAction instance. @throws Error if no action type has been set. */
  build(): AdaptiveAction {
    if (!this.action) {
      throw new Error(
        'No action type specified. Call openUrl(), submit(), showCard(), toggleVisibility(), or execute() first.',
      );
    }
    return this.action;
  }
}
